### PR TITLE
(Feature) Unique Input Async Validator

### PR DESCRIPTION
**Pull request by sfount**
This pull request implements https://github.com/IMA-WorldHealth/bhima-2.X/issues/220.  
### Example (taken from source documentation)

``` js
@example 
<input id="email" type="email" bh-unique="/users/validateEmail">
```

`bh-unique` - the server end point that will verify if the value passed (through the input model) is unique 

The server route should implement a very simple API that returns either true or false given the input model:
`GET /example/checkInputValue/:value`
`true` - the value has been found
`false` - the value has not been found
### Code reference

``` html
                    <input 
                    id="new-hospital-number"
                    name="hospitalNumber"
                    class="form-control"
                    autocomplete="off"
                    ng-model="PatientRegCtrl.medical.hospital_no"
                    bh-unique="/patients/hospital_number"
                    required>

                  <!-- This is entirely optional - it drives the in-line feedback icons on the input -->
                  <span ng-if="details.hospitalNumber.$error.unique" class="glyphicon glyphicon-remove form-control-feedback" id="unique-error-icon"></span>
                  <span ng-if="details.hospitalNumber.$pending.unique"  class="glyphicon glyphicon-hourglass form-control-feedback"></span>

                  <!-- ng-message feedback -->
                  <div class="help-block" ng-messages="details.hospitalNumber.$error" ng-show="details.$submitted || details.hospitalNumber.$error.unique">
                    <p ng-messages-include="partials/templates/messages.tmpl.html"></p> 
                  </div>
```

Closes https://github.com/IMA-WorldHealth/bhima-2.X/issues/220

---

Hi! Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](./docs/STYLEGUIDE.md)
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub)
that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process
and help build a stronger application.  Thanks!

### CODE REVIEW

**Comment by sfount** [comment]
@jniles @DedrickEnc If you have a chance please look over the code. I will perform a formal code review as well. 

@jniles If this meets your issues requirements this can be merged.

File: client/src/js/directives/bhUnique.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+/** 
+ * Unique Input Directive 
+ *
+ * This input extends an input to add an asynchronous unique constraint. The 
+ * directive will make HTTP requests to a specified URL and expect to recieve 
+ * `true` and `false` values. This will then update the $valid property 
+ * on the angluar form record. 
+ *
+ * @example 
+ * <input id="email" type="email" bh-uniqe="/users/validateEmail">
```
Comment (jniles) [inline client/src/js/directives/bhUnique.js]: Important typo: `bh-uniqe` --> `bh-unique`.

**Comment by jniles** [comment]
@sfount, this looks like it will meet my needs and the code is clean.  If you could re-implement the hospital unique lookup with this, it would shed some light on what the server-side routes must look like and provided a good test case for this directive.

**Comment by sfount** [comment]
@jniles thanks for taking the time to look at that. 

Reference - [implemented the hospital number lookup](https://github.com/IMA-WorldHealth/bhima-2.X/pull/235/commits/21659bf4097fa6d5fb7ace1b09132b0e94959c01) and [ensured the server side routes were tested](https://github.com/IMA-WorldHealth/bhima-2.X/pull/235/commits/8b392b938f9f1a42754522c5c63d36165d360dbb). As all end to end tests it proves that functionality is on par with the previous implementation. Your issue https://github.com/IMA-WorldHealth/bhima-2.X/issues/234 will help with testing directives and components and should be set as high priority. 

``` html
                    <input 
                    id="new-hospital-number"
                    name="hospitalNumber"
                    class="form-control"
                    autocomplete="off"
                    ng-model="PatientRegCtrl.medical.hospital_no"
                    bh-unique="/patients/hospital_number"
                    required>

                  <!-- This is entirely optional - it drives the in-line feedback icons on the input -->
                  <span ng-if="details.hospitalNumber.$error.unique" class="glyphicon glyphicon-remove form-control-feedback" id="unique-error-icon"></span>
                  <span ng-if="details.hospitalNumber.$pending.unique"  class="glyphicon glyphicon-hourglass form-control-feedback"></span>

                  <!-- ng-message feedback -->
                  <div class="help-block" ng-messages="details.hospitalNumber.$error" ng-show="details.$submitted || details.hospitalNumber.$error.unique">
                    <p ng-messages-include="partials/templates/messages.tmpl.html"></p> 
                  </div>
```

**Comment by jniles** [comment]
@sfount, this is a very nice, clean implementation.  The API tests are a huge help, and add a lot of clarity.  Thanks!

From a functionality perspective, LGTM.  Satisfies #220!

**Comment by sfount** [comment]
Thanks for the quick feedback @jniles

**Comment by sfount** [comment]
This code has [passed a code review](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) as per my initial review.

If anyone has any additional comments or problems with this pull request it can be undone and reviewed.

**Pull request merged**